[quant][graphmode][fx] Add support for weight and bias dtype in backend_config_dict

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68229
* #68228
* #68227
* #68042
* #67537
* __->__ #68602

Summary:
This PR adds support for configuring weight/bias dtype in backend_config_dict
and refactor the current code that checks when to insert observers

Test Plan:
Regression
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Actual test of functionality will be in next PR
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537712](https://our.internmc.facebook.com/intern/diff/D32537712)

jerryzh168 committed:
[quant][graphmode][fx] Add support for weight and bias dtype in backend_config_dict

Summary:
This PR adds support for configuring weight/bias dtype in backend_config_dict
and refactor the current code that checks when to insert observers

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537712).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for weight and bias dtype in backend_config_dict"


Summary:
This PR adds support for configuring weight/bias dtype in backend_config_dict
and refactor the current code that checks when to insert observers

Test Plan:
Regression
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Actual test of functionality will be in next PR
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537712](https://our.internmc.facebook.com/intern/diff/D32537712)

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -372,6 +367,27 @@ def get_arg_target_dtype(
     else:
         return node_name_to_target_dtype[arg.name]["output_activation_dtype"]
 
+def get_arg_target_dtype_as_input(
```
sorry, what does this mean conceptually?  Is `get_arg_target_dtype_as_output` getting the output dtype of arg, and `get_arg_target_dtype_as_input` getting the input dtype of arg?  Or something else?

vkuzo replied:
after reading the code I get it...how about something like
`get_target_dtype_of_arg_from_previous_node` and `get_target_dtype_of_arg_in_cur_node`?

vkuzo replied:
or maybe `get_arg_target_dtype_as_output` and `get_arg_target_dtype_as_input_to_this_node`?

jerryzh168 replied:
yeah I feel this is a bit hard to name as well, I will use something like `get_arg_target_dtype_as_output` and `get_arg_target_dtype_as_input_to_this_node`

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for weight and bias dtype in backend_config_dict"


Summary:
This PR adds support for configuring weight/bias dtype in backend_config_dict
and refactor the current code that checks when to insert observers

Test Plan:
Regression
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Actual test of functionality will be in next PR
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537712](https://our.internmc.facebook.com/intern/diff/D32537712)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537712).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for weight and bias dtype in backend_config_dict"


Summary:
This PR adds support for configuring weight/bias dtype in backend_config_dict
and refactor the current code that checks when to insert observers

Test Plan:
Regression
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Actual test of functionality will be in next PR
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537712](https://our.internmc.facebook.com/intern/diff/D32537712)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537712).
